File: Number of People Aware of a Secret/Number of People Aware of a Secret.py
# Solution for Number of People Aware of a Secret in PY

class Solution:
    MOD = 10**9 + 7

    def countNewPeople(self, day: int, delay: int, forget: int, memo: list) -> int:
        if day == 1:
            return 1  # Base case: Day 1 → 1 person knows the secret

        if memo[day] != -1:
            return memo[day]  # Already computed

        ways = 0
        # Loop through possible "starting days"
        for d in range(day - forget + 1, day - delay + 1):
            if d > 0:
                ways = (ways + self.countNewPeople(d, delay, forget, memo)) % self.MOD

        memo[day] = ways
        return ways

    def peopleAwareOfSecret(self, n: int, delay: int, forget: int) -> int:
        memo = [-1] * (n + 1)
        total = 0

        # Only people who haven't forgotten remain in the last window
        for day in range(n - forget + 1, n + 1):
            if day > 0:
                total = (total + self.countNewPeople(day, delay, forget, memo)) % self.MOD

        return total

File: Number of People Aware of a Secret/Number of People Aware of a Secret.cpp
// Solution for Number of People Aware of a Secret in CPP

class Solution {
public:
    const int MOD = 1e9 + 7;

    // Recursive helper function:
    // Returns the number of new people who learn the secret on 'day'
    int countNewPeople(int day, int delay, int forget, vector<int> &memo) {
        if (day == 1) return 1; // Base case: Day 1, exactly 1 person knows the secret

        if (memo[day] != -1) return memo[day]; // Already computed → return memoized value

        int ways = 0;

        // Loop through all possible "starting days" d such that
        // someone who learned on 'd' can pass the secret to 'day'
        for (int d = day - forget + 1; d <= day - delay; d++) {
            if (d > 0) { 
                ways = (ways + countNewPeople(d, delay, forget, memo)) % MOD;
            }
        }

        return memo[day] = ways;
    }

    int peopleAwareOfSecret(int n, int delay, int forget) {
        vector<int> memo(n + 1, -1); // Memoization table for days
        int total = 0;

        // At the end of day n, only people who haven’t forgotten remain.
        // That means they learned the secret in the window [n - forget + 1, n].
        for (int day = n - forget + 1; day <= n; day++) {
            if (day > 0) {
                total = (total + countNewPeople(day, delay, forget, memo)) % MOD;
            }
        }

        return total;
    }
};
